src/librustc_lint/builtin.rs

```diff
@@ -673,6 +673,75 @@ impl EarlyLintPass for AnonymousParameters {
     }
 }
 
+/// Checks for incorrect use use of `repr` attributes.
+#[derive(Clone)]
+pub struct BadRepr;
+
+impl LintPass for BadRepr {
+    fn get_lints(&self) -> LintArray {
+        lint_array!()
+    }
+}
+
+impl EarlyLintPass for BadRepr {
+    fn check_attribute(&mut self, cx: &EarlyContext, attr: &ast::Attribute) {
+        if attr.name() == "repr" {
+            let list = attr.meta_item_list();
+
+            // Emit warnings with `repr` either has a literal assignment (`#[repr = "C"]`) or
+            // no hints (``#[repr]`)
+            let has_hints = list.as_ref().map(|ref list| !list.is_empty()).unwrap_or(false);
+            if !has_hints {
+                let mut suggested = false;
+                let mut warn = if let Some(ref lit) = attr.value_str() {
+                    // avoid warning about empty `repr` on `#[repr = "foo"]`
+                    let sp = match format!("{}", lit).as_ref() {
+                        "C" | "packed" | "rust" | "u*" | "i*" | "transparent" => {
+                            let lo = attr.span.lo() + BytePos(2);
+                            let hi = attr.span.hi() - BytePos(1);
```

Comment:
It's to be 100% confident that the suggested code will be correct like in the following case:

```
warning: `repr` attribute isn't configurable with a literal
 --> file.rs:1:3
  |
1 | #[   repr   =    "C"   ]
  |   ^^^^^^^^^^^^^^^^^^^^^ help: give `repr` a hint: `repr(C)`
```

In the cases where we're not suggesting anything `attr.span` is good enough. If we use `attr.span` for the case above, the output would be:

```
warning: `repr` attribute isn't configurable with a literal
 --> file.rs:1:3
  |
1 | #[   repr   =    "C"   ]
  |   ^^-------------------^
  |     |
  |     help: give `repr` a hint: `repr(C)`
```

Reply:
I don't really follow. I guess it might explain why these calculations are only done in one branch, but why are they done at all? Why try to exclude the `#[` and `]` from the labels in the first place? It doesn't significantly change how the diagnostic looks and I haven't seen other attribute-related diagnostics do it.

Reply:
I though that `repr` attrs could be applied to the enclosing element, like other attributes `#![repr(C)]`, but I just tried it and it was rejected. Is this the case everywhere? In that case I don't have to worry about the possible difference between `#![repr]` and `#[repr]` and wouldn't have to do this span wrangling.

Reply:
I don't think `#![repr]` is valid, but I also don't get why that would make any difference for the span. The span points at the attribute either way, doesn't it? (Actually, I am now even more confused: if `#![repr]` *was* valid, `lo + 2` would be off by one for it)